compiler/packages/react-mcp-server/src/utils/runtimePerf.ts

```diff
@@ -0,0 +1,138 @@
+import * as babel from '@babel/core';
+import * as parser from '@babel/parser';
+import puppeteer from 'puppeteer';
+
+export async function measurePerformance(code: any) {
+    // Parse the code into an AST
+    const parsed = await parseAsync(code);
+
+    // Transform the AST into browser-compatible JavaScript
+    const transpiled = await transformAsync(parsed, code);
+
+    console.log(transpiled);
+
+    // Launch puppeteer with increased protocol timeout
+    const browser = await puppeteer.launch({
+        protocolTimeout: 600_000, // Increase timeout to 10 minutes
+        headless: false
+    });
+    const page = await browser.newPage();
+    await page.setViewport({ width: 1280, height: 720 });
+    const html = buildHtml(transpiled);
+    await page.setContent(html, { waitUntil: 'networkidle0' });
+    await page.waitForFunction('window.__RESULT__ !== undefined', {timeout: 600_000}); // 10 minute timeout
+
+    const result = await page.evaluate(() => {
+        return (window as any).__RESULT__;
+    });
+    // await browser.close();
+    return result;
+}
+
+/**
+ * Parse code string into an AST
+ * @param {string} code - The source code to parse
+ * @returns {babel.types.File} - The parsed AST
+ */
+function parseAsync(code: string) {
+    return parser.parse(code, {
+        sourceType: 'module',
+        plugins: [
+            'jsx',
+            'typescript',
+            'classProperties',
+            'optionalChaining',
+            'nullishCoalescingOperator',
+        ],
+    });
+}
+
+/**
+ * Transform AST into browser-compatible JavaScript
+ * @param {babel.types.File} ast - The AST to transform
+ * @param {Object} opts - Transformation options
+ * @returns {Promise<string>} - The transpiled code
+ */
+async function transformAsync(ast: babel.types.Node, code: string) {
+    // Provide a dummy filename to satisfy Babel's requirement for filename context
+    const result = await babel.transformFromAstAsync(ast, null,  {
+        filename: 'file.jsx',
+        presets: [
+          ['@babel/preset-env'],
+          '@babel/preset-react'
+        ],
+        plugins: [
+            () => ({
+                visitor: {
+                  ImportDeclaration(path) {
+                    const value = path.node.source.value;
+                    if (value === 'react' || value === 'react-dom') {
+                      path.remove();
+                    }
+                  }
+                }
+              })
+        ]
+    });
+
+    return result?.code || '';
+}
+
+function buildHtml(transpiled: string | null | undefined) {
```

Comment:
This argument really should be non-nullable. There is currently no handling of the null/undefined case here. I would move this check into the callsite instead and throw early if `transpiled` is ever null/undefined since there's nothing to do if the code couldn't be transpiled.